### ISSUE TITLE
Calculating zenith of a star

### ISSUE THREAD

**Issue by baisong** [cosinekitty/astronomy#221]
Hello & thank you for this phenomenal tool!

I want to use the Javascript library to determine the zenith of the star [Alnilam](https://www.wikiwand.com/en/Alnilam), aka [ε](https://www.wikiwand.com/en/Bayer_designation) Ori, [46](https://www.wikiwand.com/en/Flamsteed_designation) Orionis, [112 G](https://www.wikiwand.com/en/Gould_designation) Orionis, [HR](https://www.wikiwand.com/en/Harvard_Revised_catalogue) 1903, [BD](https://www.wikiwand.com/en/Bonner_Durchmusterung) -01°969, [HD](https://www.wikiwand.com/en/Henry_Draper_Catalogue) 37128, [SAO](https://www.wikiwand.com/en/Smithsonian_Astrophysical_Observatory_Star_Catalog) 132346, [FK5](https://www.wikiwand.com/en/FK5) 210, [HIP](https://www.wikiwand.com/en/Hipparcos_catalogue) 26311, TD1 4963, [参宿二](https://www.wikiwand.com/en/Three_Stars_(Chinese_constellation)), which, according to traditional knowledge, should be visible from San Diego, California around the time of the southward/autumn equinox.

I found [this code snippet](https://github.com/cosinekitty/astronomy/blob/ab32f548e806221f28152dd8cfd908887d57b867/generate/novas/example.c#L91) and [it's function definition](https://github.com/cosinekitty/astronomy/blob/e268637828fbb5626b49cbbfe8ff4df0c763e610/generate/novas/novas.h#L507) in one of the library files that seems to relate to this task:

```c
make_cat_entry ("GMB 1830","FK6",1307,11.88299133,37.71867646, 
      4003.27,-5815.07,109.21,-98.8, &star);
```

```c
   short int make_cat_entry (char star_name[SIZE_OF_OBJ_NAME],
                             char catalog[SIZE_OF_CAT_NAME],
                        long int star_num, double ra, double dec,
                        double pm_ra, double pm_dec, double parallax,
                        double rad_vel,

                        cat_entry *star);
```

Are there equivalent features in the Javascript library? How can I find the zenith for a given gregorian year?

**Comment by cosinekitty** [comment]
Hi @baisong and welcome! I'm not clear on what your goal is. Can you explain more about what you mean by finding the zenith of a star? When I hear the word zenith, I think about the point that is straight up from where a particular observer is standing. Can you give an example of what the answer would look like?

**Comment by cosinekitty** [comment]
I did a little more reading, and it looks like you might be asking about finding the time when the star crosses the observer's meridian. In other words, finding the time when the star reaches the highest point in the sky. Is that what you are looking for?

**Comment by baisong** [comment]
Yes, thank you for helping me clarify, according to [this modern adaptation of the Kumeyaay Calendar](http://www.kumeyaay.info/calendar/kumeyaay_calendar.html):

> "Certain celestial events helped to readjust and reset the year. The most important of these events was the ascendency of \[Orion's Belt\] to its zenith \[sic\] just before dawn. This corresponds to the three stars of Orion's belt and they reach their zenith around the fall equinox of September 21st."

**Comment by cosinekitty** [comment]
Here is a function that should help. I call the function using the J2000 equatorial coordinates of Anilam, for an observer in San Diego.

```javascript
const Astronomy = require('./astronomy.js');

function FindNextCulmination(searchTime, observer, ra_eqj, dec_eqj) {
    // Convert J2000 angular coordinates to a vector.
    const sphere_eqj = new Astronomy.Spherical(dec_eqj, 15*ra_eqj, 1);
    const vec_eqj = Astronomy.VectorFromSphere(sphere_eqj, searchTime);

    // Convert J2000 coordinates (EQJ) to of-date coordinates (EQD).
    const rot = Astronomy.Rotation_EQJ_EQD(searchTime);
    const vec_eqd = Astronomy.RotateVector(rot, vec_eqj);

    // Find of-date angular coordinates from vector.
    const sphere_eqd = Astronomy.SphereFromVector(vec_eqd);
    const ra_eqd = sphere_eqd.lon / 15;     // convert degrees to sidereal hours

    // Calculate how much time it will take to bring the star's right ascension
    // to the observer's meridian.
    const st = Astronomy.SiderealTime(searchTime);
    let delta_sidereal_hours = ((ra_eqd - observer.longitude/15) - st) % 24;
    if (delta_sidereal_hours < 0) {
        delta_sidereal_hours += 24;     // always search forward in time
    }

    const SOLAR_DAYS_PER_SIDEREAL_DAY = 0.9972695717592592;
    const delta_days = (delta_sidereal_hours / 24) * SOLAR_DAYS_PER_SIDEREAL_DAY;
    const time = searchTime.AddDays(delta_days);
    console.log(`Culmination at: ${time}`);
}


function Demo() {
    // https://www.wikiwand.com/en/Alnilam
    // The J2000 coordinates of Anilam:
    const ra_eqj  =   5 + 36/60 + 12.8/3600;
    const dec_eqj = -(1 + 12/60 +  6.9/3600);

    // Coordinates for San Diego, California.
    const observer = new Astronomy.Observer(32.82, -117.11, 0);

    // Define a time after which to find the next star culmination.
    const searchTime = Astronomy.MakeTime(new Date('2022-05-16T00:00:00Z'));
    FindNextCulmination(searchTime, observer, ra_eqj, dec_eqj);
}

Demo();
```

Here is the output:

```
$ node ./starculm.js
Culmination at: 2022-05-16T21:47:26.412Z
```

The main complexity here is that star coordinates are usually given with respect to the Earth's equator on January 1, 2000. This is the so-called J2000 epoch.  You have to correct for changes in the Earth's axis (precession and nutation) for the given date, in order to know where the star actually appears in the sky.  The above function `FindNextCulmination` does that. Then it does a sidereal time calculation to figure out when the star crosses the meridian for the given observer's longitude.

Please let me know if this makes sense, and whether it works for what you are trying to do.

**Comment by baisong** [comment]
Thank you so much for this, @cosinekitty!!

I have this working locally and will play around a bit over the next week. I think the only other information we will need is how to calculate the time of dawn on a given date from the same observer's perspective...

Like this?
```js
function FindDawnFromSanDiego(dateStart)
    // Coordinates for San Diego, California.
    const observer = new Astronomy.Observer(32.82, -117.11, 0);
    return SearchAltitude("Sun", observer, 1, dateStart, 1, 0);
}
```

**Comment by cosinekitty** [comment]
That's great!

There is a better way to find sunrise that also takes into account atmospheric refraction and the fact that sunrise is based on the apparent edge of the Sun, not its center.  See: [SearchRiseSet](https://github.com/cosinekitty/astronomy/tree/master/source/js#SearchRiseSet). SearchAltitude is better for finding things like civil twilight or nautical twilight.
